index.bs

```diff
@@ -2015,14 +1991,76 @@ Issue: This needs to be generalised to work with realms too.
 <div algorithm="remote end steps for session.unsubscribe">
 The [=remote end steps=] with |session| and |command parameters| are:
 
-1. Let the |list of event names| be the value of the <code>events</code> field of
-   |command parameters|.
+1. If |command parameters| matches the <code>session.UnsubscribeByAttributesRequest</code> production:
+
+   1. Let |event names| be a [=/set=] of |command parameters|["<code>events</code>"].
+
+   1. Let |contexts| be a [=/set=] of |command parameters|["<code>contexts</code>"],
+      if it is present or an empty [=/set=] if it isn't.
+
+   1. Let |new subscriptions| to be a [=/list=].
+
+   1. For each |subscription| of |session|'s [=subscriptions=]:
+
+      1. If [=set/intersection=] of |subscription|'s [=subscription/event names=] and |event names| is an empty [=/set=],
+         continue.
+
+      1. If |contexts| is an empty [=/set=]:
+
+        1. |=list/Remove=| all |event names| from |subscription|'s [=subscription/event names=].
+
+        1. If |subscription|'s [=subscription/event names=] is not empty:
+
+          1. [=list/append=] |subscription| to |new subscriptions|.
+
+      1. Otherwise:
+
+        1. If |subscription| is [=subscription/global=]:
+
+          1. [=list/append=] |subscription| to |new subscriptions|.
+
+        1. Otherwise:
+
+          Note: unsubscribe by contexts is deprecated and will be removed in the future versions.
+
+          1. Let |event map| be an empty [=/map=].
+
+          1. For each |event name| in |subscription|'s [=subscription/event names=]:
+
+              1. Set |event map|[|event name|] to [=set/clone=] of |subscription|'s [=subscription/contexts=].
+
+          1. For each |event name| in |event names|:
+
+              1. If |event map|[|event name|] does not exist,
+                 continue.
+
+              1. Set |event map|[|event name|] to the [=set/difference=]
+                 between |event map|[|event name|] and |contexts|.
+
+              1. If |event map|[|event name|] is an empty [=/set=],
+
+                1. [=map/Remove=] |event map|[|event name|].
+
+          1. For each |event name| → |remaining contexts| in |event map|:
+
+            1. Let |subscription| be a [=subscription=] with
+               [=subscription/subscription id=] to |subscription|'s [=subscription/subscription id=],
+               [=subscription/event names=] set to a new [=/set=] containing |event name|,
+               [=subscription/contexts=] set to |remaining contexts|.
+
+   1. Set |session|'s [=subscriptions=] to |new subscriptions|.
+
+1. Otherwise:
+
+   1. Let |subscriptions| be a [=/set=] of |command parameters|[<code>subscriptions</code>].
+
+   1. Let |items to remove| be an empty [=/list=].
 
-1. Let the |list of contexts| be the value of the <code>contexts</code>
-   field of |command parameters| if it is present or null if it isn't.
+   1. For each |subscription| in |session|'s [=subscriptions=],
```

Comment:
So if we try to unsubscribe with a subscription id that doesn't exist we just silently ignore it. That isn't consistent with `script.removePreloadScript` which will error in that case. I can see the case for ignoring, but given the precedent I'd prefer we error if any invalid subscription ids are passed in.

Reply:
Let's say we would remove subscriptions if all related contexts were destroyed, should we still error in that situation?

Reply:
No, I think the rule should be that if something was a valid subscription that hasn't been removed it's OK to remove, even if events in the meantime have rendered it a no-op. That isn't suggestive of a client bug, whereas trying to remove a subscription that was never created, or has already been removed, is.

Reply:
I have updated the spec to throw an error. I think one edge case to consider is whether we want to throw an error if removal by ID finds no subscriptions but unsubscribe happened previously by attributes (the current spec text would throw).